### PR TITLE
Add Control Plane Connector Commands

### PR DESCRIPTION
**Pull request by saterus**
**Description:**

Adds endpoints which query connectors directly for information. Primarily:
* `GET /connector_images/:image_id/spec`
* `POST /connector_images/:image_id/discovery`

We've added structure to the payload formats so we can consistently format our payloads across all the endpoints. Each payload contains self-discoverable links inside the payloads to make traversing the API easy and contextual to the API consumer.

This PR also adds CORS support so the UI can interact with these endpoints from the browser.

This adds a dependency on `flowctl`, since these endpoints shell out for information. This complicates the old build process, and necessitated overhauling the build.

To add to the complexity here, we've circularly made `flowctl-rs` itself depend on `control`. This moves all of the old `control/src/bins/*` into `flowctl-rs` subcommands. Now users can manage the control plane with:
* `flowctl control-plane setup`: Create and/or migrate the control plane database
* `flowctl control-plane seed`: Seed the control plane database
* `flowctl control-plane serve`: Launch a control plane locally
* `flowctl control-plane id-mask`: Encode/decode entity ids for debugging purposes. This command probably won't live here long term, but it is useful in development.

**Workflow steps:**

*Start the server*:
```
$ cd path/to/flow/crates/control
$ flowctl-rs control-plane setup --config=config/development.yaml
$ flowctl-rs control-plane seed --config=config/development.yaml
$ flowctl-rs control-plane serve --config=config/development.yaml

# Alternatively, for active development on the control plane itself, you can forgo `flowctl-rs`.
$ sqlx database setup
$ cargo run
```

*Fetch a spec*:

```bash
# Find the link to the spec for the postgres connector. Your IDs will vary.
$ curl localhost:3000/connector_images | jq '.data[] | select(.attributes.name | contains("postgres") ) | .links.spec'
"http://127.0.0.1:3000/connector_images/JoWaIxGABAQ/spec"

# Fetch the connector spec
$ curl http://127.0.0.1:3000/connector_images/JoWaIxGABAQ/spec
```

*Discovery bindings*:

```bash
# Find the link to the discovery endpoint:
$ curl http://127.0.0.1:3000/connector_images/JoWaIxGABAQ/spec | jq '.data.links.discovery'
"http://127.0.0.1:3000/connector_images/JoWaIxGABAQ/discovery"

# Perform discovery on this source
$ curl -XPOST -H "Content-Type: application/json" -d @- http://127.0.0.1:3000/connector_images/JoWaIxGABAQ/discovery
{
  "database": "control_development",
  "host": "127.0.0.1",
  "password": "flow",
  "port": 5432,
  "user": "flow"
}
```

**Documentation links affected:**

Now that we've added commands to flowctl-rs, we should probably begin thinking about how much documentation to add. I'm unsure at the moment, as we've said the Control Plane UI is taking precedence over the CLI for users. I've added small bits of documentation to the CLI actions, but nothing to the main docs site yet.

**Notes for reviewers:**

There's a lot in this PR, often unrelated to other bits. I'm so sorry. Hopefully once we work the build issues out, we'll never do this again. Commit-by-commit is your only chance at survival.

---
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/376)

### CODE REVIEW

File: crates/control/.cargo/config [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Match build args from the workspace cargo config
+[build]
+rustflags = ["-C", "link-arg=-fuse-ld=lld"]
```
Comment (psFried) [inline crates/control/.cargo/config:3]: Just making sure I understand why this is needed. Does this file completely override the top level .cargo/config?

**Review by psFried — APPROVED** [review]
This is looking great! I had some questions, but I don't think any of them are necessarily blocking.

File: crates/control/src/services/connectors.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+use tokio::process::Command;
+
+pub fn spec(image: &str) -> Command {
+    let mut cmd = Command::new("flowctl");
+    cmd.arg("api")
+        .arg("spec")
+        .arg("--image")
+        .arg(image)
+        .arg("--network")
+        .arg("host");
```
[the remark this answers is not in the document]
Comment (saterus) [reply crates/control/src/services/connectors.rs:12]: Good catch. :facepalm:

File: crates/control/src/services/connectors.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+pub fn discovery(image: &str, config_path: &Path) -> Command {
+    let mut cmd = Command::new("flowctl");
+    cmd.arg("api")
+        .arg("discover")
+        .arg("--image")
+        .arg(image)
+        .arg("--config")
+        .arg(config_path)
+        .arg("--network")
```
[the remark this answers is not in the document]
Comment (saterus) [reply crates/control/src/services/connectors.rs:25]: Good idea. I can do that pretty easily as part of this PR.

File: crates/control/src/controllers/connector_images.rs [hunk 4]
```diff
@@ -21,5 +27,48 @@ pub async fn create(
...
+    Path(image_id): Path<Id>,
+    Json(input): Json<RawJson>,
+) -> Result<impl IntoResponse, AppError> {
+    let image = images_repo::fetch_one(&db, image_id).await?;
+
+    // TODO: We should probably allow `flowctl api discover` to take this from
+    // stdin so we don't have to write the config to a file. This is unnecessary
+    // and tempfile::NamedTempFile does not _guarantee_ cleanup, which is not
+    // great considering it may include credentials.
+    let tmpfile = tempfile::NamedTempFile::new()?;
```
[the remark this answers is not in the document]
Comment (saterus) [reply crates/control/src/controllers/connector_images.rs:65]: Good idea. I'm thinking I'll wait on this one until I revisit the discovery endpoint more generally. This is going to significantly change with the sops work (not that the named pipe isn't a better plan).

File: crates/control/src/controllers/connector_images.rs [hunk 5]
```diff
@@ -1,18 +1,24 @@
...
-use crate::controllers::Payload;
+use crate::controllers::json_api::RawJson;
 use crate::error::AppError;
 use crate::models::connector_images::CreateConnectorImage;
+use crate::models::Id;
 use crate::repo::connector_images as images_repo;
+use crate::services::connectors;
+use crate::services::subprocess::Subprocess;
+
+pub mod routes;
```
[the remark this answers is not in the document]
Comment (saterus) [reply crates/control/src/controllers/connector_images.rs:15]: Yeah, `links` is a definitely better name for what I've got here. In Rails/Phoenix, these route helpers *are* generated by the central Router. This was kind of a stopgap since I didn't have a plan for doing that.

However, things move quickly. *After* I added these `routes` modules, a plan for [Type safe routing](https://github.com/tokio-rs/axum/pull/756) was formulated and merged into Axum. This morning they released a new version that contains this work. I'll be revisiting all the route/link generation stuff as part of that upgrade.

File: crates/control/.env [hunk 6]
```diff
@@ -1 +1,8 @@
+# Use direnv or similar tool to load this into the environment automatically.
+# https://github.com/direnv/direnv
+
+# SQLx requires that the DATABASE_URL be provided via the environment in order
```
[the remark this answers is not in the document]
Comment (saterus) [reply crates/control/.env:4]: I'm going to have to shake my fist at SQLx again for this. It knows how to read the `.env` file or from the actual environment, but apparently whatever trickery Cargo uses doesn't work. Even when I run `cargo sqlx migrate run`, it complains about missing a `DATABASE_URL`. This was a bit surprising to me, but unfortunately it doesn't solve having it defined here.

My intention at this point is for developers (`cargo build/test/run`) to need this set when they are actively changing/compiling the project, but that users (`flowctl-rs control-plane`) will always be passing a path to a config file that contains the database settings. In production, we can inject the entire config file as a Secret, or individual Secrets that get merged into a stock config file.

File: crates/control/.cargo/config (hunk 1 above)
Comment (saterus) [reply crates/control/.cargo/config:3]: I added this when I thought it was the source of recompilation issues in CI. I *think* that Cargo looks for a config file in the hierarchy, but isn't necessarily merging them together. When I add/remove this line and run the build from inside `crates/control`, it rebuilds all the dependencies. :grimacing: To double check this, I did a `cargo clean` and then `cargo build` within `crates/control` with and then without these lines. Afterwards I end up with `target/debug` and `target/x86_64-unknown-linux-gnu/debug`. :sob: 

In summary, this is another sharp edge of workspaces. If you are working at the workspace root and also within a specific crate, you'll end up getting different compilation artifacts.

File: crates/control/src/controllers/connector_images.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (psFried) [reply crates/control/src/controllers/connector_images.rs:65]: :+1:

File: crates/control/src/controllers/connector_images.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (psFried) [reply crates/control/src/controllers/connector_images.rs:15]: neat! Yeah, at this point it makes sense to just look into switching to the type safe routing instead of fiddling with that ourselves.

File: crates/control/.env (hunk 6 above)
[the remark this answers is not in the document]
Comment (psFried) [reply crates/control/.env:4]: Ok, I kind of expected there was some reason for that.
